File: periodo_2/STL/unidade_1/projetos/sistemadeimpressao/danilonascimento_202000137609_sistemadeimpressao.go
// Package main provides ...
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

type Lista struct {
	element Documento
	next *Lista
}

type Fila struct {
	head   *Lista
	tail   *Lista
	length int
}

type Pilha struct {
	head   *Lista
	tail   *Lista
	length int
}

type Documento struct {
	nome    string
	paginas int
}

type Impressora struct {
	nome       string
	documentos Pilha
	impressoes int
}

type Sistema_de_impressao struct {
	impressoras []Impressora
	documentos  Pilha
	impressoes  int
}

// ******* Interface para fila de documentos ******* {{
func (fila *Fila) add(element Documento) {
	if fila.length == 0 {
		fila.head = &Lista{element: element, next: nil}
		fila.tail = fila.head
	} else {
		fila.head = &Lista{element: element, next: fila.head}
	}
	fila.length++
}

func (fila *Fila) pop() (element Documento) {
	current := fila.head
	if fila.length > 0 {
		if fila.length == 1 {
			element = fila.head.element
			fila.head = nil
		} else {
			for current.next.next != nil {
				current = current.next
			}
			element = current.next.element
			current.next = nil
		}
		fila.length--
	}
	return
}

func (fila *Fila) show() {
	var current *Lista = fila.head

	for current != nil {
		fmt.Println(current.element)
		current = current.next
	}
}

// }}

// ******* Interface para pilha de documentos ******* {{
func (pilha *Pilha) add(element Documento) {
	if pilha.length == 0 {
		pilha.head = &Lista{element: element, next: nil}
		pilha.tail = pilha.head
	} else {
		pilha.head = &Lista{element: element, next: pilha.head}
	}
	pilha.length++
}

func (pilha *Pilha) pop() (element Documento) {
	if pilha.length > 0 {
		element = pilha.head.element
		if pilha.length == 1 {
			pilha.head = nil
		} else {
			pilha.head = pilha.head.next
		}
		pilha.length--
	}
	return
}
// }}
/*
	O tempo para imprimir as paginas dos documentos são contantes
	com isso, é definido qual impressora realizara a impressão com base
	na quatidade paginas que elas ja imprimiu, logo a proxima impressora
	a imprimir é a que tiver menos impressões
	impressora. documentos < ...Impressooras
*/
func (sistema_de_impressao *Sistema_de_impressao) imprimir(documento Documento) string {
	proxima := &Impressora{nome: "", impressoes: 999999}

	for i := 0; i < len((*sistema_de_impressao).impressoras); i++ {
		if proxima.impressoes > (*sistema_de_impressao).impressoras[i].impressoes {
			proxima = &(*sistema_de_impressao).impressoras[i]
		}
	}

	if proxima.documentos.head != nil {
		sistema_de_impressao.documentos.add(proxima.documentos.head.element)
	}
	sistema_de_impressao.impressoes += documento.paginas
	proxima.imprimmir(documento)
	return proxima.show()
}

func (impressora *Impressora) imprimmir(documento Documento) {
	impressora.documentos.add(documento)
	impressora.impressoes += documento.paginas
}

func (impressora Impressora) show() (show string) {
	show = fmt.Sprintf("[%s] ", impressora.nome)
	current := impressora.documentos.head
	for current != nil {
		if current.next != nil {
			show += fmt.Sprintf("%s-%dp, ", current.element.nome, current.element.paginas)
		} else {
			show += fmt.Sprintf("%s-%dp", current.element.nome, current.element.paginas)
		}
		current = current.next
	}
	return
}

func read_file(file_name string) ([]string, error) {
	arquivo, err := os.Open(file_name)

	if err != nil {
		return nil, err
	}

	defer arquivo.Close()

	var linhas []string
	scanner := bufio.NewScanner(arquivo)

	for scanner.Scan() {
		linhas = append(linhas, scanner.Text())
	}

	return linhas, scanner.Err()
}

func write_file(file *os.File, message string) {
	file.Seek(0, 2)
	file.WriteString(fmt.Sprintln(message))
	file.Sync()
}

func async_write_file(file *os.File, message string) {
	file.Seek(0, 2)
	file.WriteString(fmt.Sprintln(message))
	go file.Sync()
}

func raise(err error) {
	if err != nil {
		log.Panic(err.Error())
	}
}

func main() {
	start := time.Now()
	var sistema_de_impressao Sistema_de_impressao

	input, err := read_file(os.Args[1])
	output_file, err_2 := os.OpenFile(os.Args[2], os.O_RDWR, 0644)

	raise(err)
	raise(err_2)

	defer output_file.Close()

	n_impressoras, err := strconv.Atoi(input[0])
	documentos := input[n_impressoras+2:]

	// Armazenando as impressoras disponiveis
	for _, value := range input[1 : n_impressoras+1] {
	sistema_de_impressao.impressoras = append(
		sistema_de_impressao.impressoras, Impressora{
											nome:       value,
											documentos: Pilha{},
											impressoes: 0,
		})
	}

	var dados []string
	var nome string
	var paginas int

	// Enfililerando todos os documentos
	for _, documento := range documentos {
		dados = strings.SplitN(documento, " ", 2)
		if len(dados) == 2 {
			nome = dados[0]
			paginas, err = strconv.Atoi(dados[1])
			raise(err)
			async_write_file(output_file, sistema_de_impressao.imprimir(Documento{nome: nome, paginas: paginas}))
		}
	}

	for _, impressora := range sistema_de_impressao.impressoras {
		if impressora.documentos.head != nil {
			sistema_de_impressao.documentos.add(impressora.documentos.head.element)
		}
	}

	async_write_file(output_file, fmt.Sprintf("%dp", sistema_de_impressao.impressoes))

	current := sistema_de_impressao.documentos.head

	for current != nil {
		async_write_file(output_file, fmt.Sprintf("%s-%dp", current.element.nome, current.element.paginas))
		current = current.next
	}
	
	fmt.Println("Time exec: ", time.Since(start))
}

File: periodo_2/STL/unidade_1/projetos/sistemadeimpressao/danilonascimento_202000137609_sistemadeimpressao.cpp
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <stdio.h>
#include <iostream>
#include <string>
#include <string.h>
#include <fstream>
#include <system_error>

using namespace std;

typedef struct Documento{
    string name;
    int paginas;
} Documento;


template <class T> class Node {
public:
  T x;
  Node *next;
  Node(T x0) : x(x0){
    next = NULL;
  }
};

template <class T> class Fila {
  string null;

protected:
  int n;

public:
  Fila() {
    n = 0;
    head = NULL;
  }

  Node<T> *head;
  Node<T> *tail;

  int size() { return n; }

  void push(T x) {
    Node<T> *u = new Node<T>(x);
    if (size() == 0) 
      head = u;
    else 
      tail->next = u;
    tail = u;
    n++;
  }

  void pop() {
      if (n > 0) {
        Node<T> *u = head;
        head = head->next;
        delete u;
        n--;
      }
  }

  void index() {
    Node<T> *u = head;

    while (u != NULL) {
        std::cout << u->x.name << std::endl;
        u = u->next;
    }

  }
};
template <class T> class Pilha {
  string null;

protected:
  int n;

public:
  Pilha() {
    n = 0;
    head = NULL;
  }

  Node<T> *head;

  int size() { return n; }

  void add(T x) {
    Node<T> *u = new Node<T>(x);
    if (size() > 0) 
      u->next = head;
    head = u;
    n++;
  }

  void pop() {
      if (n > 0) {
        Node<T> *u = head;
        head = head->next;
        delete u;
        n--;
      }
  }

  void index() {
    Node<T> *u = head;

    while (u != NULL) {
        std::cout << u->x.name << std::endl;
        u = u->next;
    }

  }
};
class Impressora {
  public:
    string name;
    int impressoes;
    Pilha<Documento> *documentos;
    Impressora(string Name) : name(Name){
      documentos = new Pilha<Documento>();
      impressoes = 0;
    };

    void imprimir(Documento documento) {
      documentos->add(documento);
      impressoes += documento.paginas;
    }

    string show() {
      string show;
      show += "["+name +"] ";
      Node<Documento> *documento = documentos->head;
      while (documento != NULL) {
        show += documento->x.name + "-" + to_string(documento->x.paginas) + "p";
        if (documento->next != NULL)
          show += ", ";
        documento = documento->next;
      }
      return show;
    } 
};

class SistemaDeImpressao 
{
public:
  SistemaDeImpressao() {
    impressoras = Fila<Impressora>();
  };
  Fila<Impressora> impressoras;
  Pilha<Documento> documentos;
  int impressoes;

  string imprimir(Documento documento) {
    Node<Impressora> *impressora = impressoras.head;
    Node<Impressora> *next = impressoras.head;

    while(impressora != NULL) {
      if (next->x.impressoes > impressora->x.impressoes)
        next = impressora;
      impressora = impressora->next;
    }

    if (next->x.documentos->head != NULL)
      documentos.add(next->x.documentos->head->x);

    impressoes += documento.paginas;
    next->x.imprimir(documento);
    return next->x.show();
  }
  Node<Impressora>* next() {
    Node<Impressora> *impressora = impressoras.head;
    Node<Impressora> *next = impressoras.head;

    while(impressora != NULL) {
      if (next->x.impressoes > impressora->x.impressoes)
        next = impressora;
      impressora = impressora->next;
    }

    return next;
  }
};

char* read_file(FILE* arg) {
  char *linha = new char[100], *result;
  result = fgets(linha, 100, arg);  // o 'fgets' lê até 99 caracteres ou até o '\n'
  if(result) 
    return linha;
  return linha;
}
int main(int argc, char* argv[]) {
  printf("#ARGS = %i\n", argc);
  printf("PROGRAMA = %s\n", argv[0]);
  printf("ARG1 = %s, ARG2 = %s\n", argv[1], argv[2]);

  SistemaDeImpressao sistema;
  char *linha;
  FILE* input = fopen(argv[1], "r");
  fstream file_out;
  file_out.open(argv[2], ios_base::out);

  if (input == NULL || !file_out.is_open()) {
     printf("Problemas com os arquivo\n");
     return EXIT_FAILURE;
  }

  linha = read_file(input);
  int n_impressoras = atoi(linha);
  delete linha;

  std::cout << n_impressoras << std::endl;
  
  for (int i = 0; i < n_impressoras; ++i) {
    linha = read_file(input);
    sistema.impressoras.push(Impressora(strtok(linha, "\n")));
    delete linha;
  }

  linha = read_file(input);
  int n_documentos = atoi(linha);
  std::cout << n_documentos << std::endl;
  delete linha;
  string name; int paginas;

  for (int i = 0; i < n_documentos; ++i) {
    linha = read_file(input);
    name = strtok (linha," ");
    paginas = atoi(strtok (NULL," "));
    delete linha;
    file_out << sistema.imprimir(Documento{name, paginas}) << std::endl;
  }

  Node<Impressora> *impressora = sistema.impressoras.head;
  
  for (int i = 0; i < n_impressoras; ++i) {
    impressora = sistema.next();
    sistema.documentos.add(impressora->x.documentos->head->x);
    impressora->x.impressoes = INT32_MAX;
  }

  Node<Documento> *documento = sistema.documentos.head;

  file_out << sistema.impressoes << "p" << endl;

  while(documento != NULL) {
    file_out << documento->x.name << "-" << documento->x.paginas << "p" << endl;
    documento = documento->next;
  }
  return EXIT_SUCCESS;
}
